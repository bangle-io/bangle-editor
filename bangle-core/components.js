import { coreMarkPlugins, coreMarkSpec } from './marks/index';
import { coreNodeSpec, coreNodePlugins } from './nodes/index';
import * as history from './extensions/history';

export * from './nodes/index';
export * from './marks/index';
export * as history from './extensions/history';

export function coreSpec(opts = {}) {
  return [
    ...coreMarkSpec(opts.node),
    ...coreNodeSpec(opts.node),
    history.spec(opts.history),
  ];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts.node),
    ...coreNodePlugins(opts.node),
    history.plugins(opts.history),
  ];
}
