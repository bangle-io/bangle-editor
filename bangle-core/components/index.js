// marks
import * as bold from './bold';
import * as code from './code';
import * as italic from './italic';
import * as link from './link';
import * as strike from './strike';
import * as underline from './underline';

// nodes
import * as doc from './doc';
import * as paragraph from './paragraph';
import * as text from './text';
import * as blockquote from './blockquote';
import * as bulletList from './bullet-list';
import * as codeBlock from './code-block';
import * as hardBreak from './hard-break';
import * as heading from './heading';
import * as horizontalRule from './horizontal-rule';
import * as listItem from './list-item/list-item';
import * as orderedList from './ordered-list';
import * as todoItem from './todo-item';
import * as todoList from './todo-list';
import * as image from './image';

// components
import * as history from './history';

// marks
export * as bold from './bold';
export * as code from './code';
export * as italic from './italic';
export * as link from './link';
export * as strike from './strike';
export * as underline from './underline';

// nodes
export * as doc from './doc';
export * as paragraph from './paragraph';
export * as text from './text';
export * as blockquote from './blockquote';
export * as bulletList from './bullet-list';
export * as codeBlock from './code-block';
export * as hardBreak from './hard-break';
export * as heading from './heading';
export * as horizontalRule from './horizontal-rule';
export * as listItem from './list-item/list-item';
export * as orderedList from './ordered-list';
export * as todoItem from './todo-item';
export * as todoList from './todo-list';
export * as image from './image';

// components
export * as history from './history';

export function coreSpec(opts = {}) {
  return [...coreMarkSpec(opts), ...coreNodeSpec(opts)];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts),
    ...coreNodePlugins(opts),
    history.plugins(opts.history),
  ];
}

function coreMarkSpec(options = {}) {
  return [
    bold.spec(options.bold),
    code.spec(options.code),
    italic.spec(options.italic),
    strike.spec(options.strike),
    link.spec(options.link),
    underline.spec(options.underline),
  ];
}

function coreMarkPlugins(options = {}) {
  return [
    bold.plugins(options.bold),
    code.plugins(options.code),
    italic.plugins(options.italic),
    strike.plugins(options.strike),
    link.plugins(options.link),
    underline.plugins(options.underline),
  ];
}

function coreNodeSpec(options = {}) {
  return [
    doc.spec(options.doc),
    text.spec(options.text),
    paragraph.spec(options.paragraph),
    blockquote.spec(options.blockquote),
    bulletList.spec(options.bulletList),
    codeBlock.spec(options.codeBlock),
    hardBreak.spec(options.hardBreak),
    heading.spec(options.heading),
    horizontalRule.spec(options.horizontalRule),
    listItem.spec(options.listItem),
    orderedList.spec(options.orderedList),
    todoItem.spec(options.todoItem),
    todoList.spec(options.todoList),
    image.spec(options.image),
  ];
}

function coreNodePlugins(options = {}) {
  return [
    paragraph.plugins(options.paragraph),
    blockquote.plugins(options.blockquote),
    bulletList.plugins(options.bulletList),
    codeBlock.plugins(options.codeBlock),
    hardBreak.plugins(options.hardBreak),
    heading.plugins(options.heading),
    horizontalRule.plugins(options.horizontalRule),
    listItem.plugins(options.listItem),
    orderedList.plugins(options.orderedList),
    todoItem.plugins(options.todoItem),
    todoList.plugins(options.todoList),
    image.plugins(options.image),
  ];
}
