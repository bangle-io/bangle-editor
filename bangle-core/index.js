import * as logging from './utils/logging';
import browser, { isChromeWithSelectionBug } from './utils/browser';
export * from './components/index';
export * from './editor';
export { logging, browser, isChromeWithSelectionBug };
export * from './pm-plugin';
export * from './node-view';
export * from './pm-model';
