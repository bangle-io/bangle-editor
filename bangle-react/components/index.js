import * as dinos from './dinos/index';

export * as dinos from './dinos/index';

export const spec = (opts = {}) => {
  return [dinos.spec(opts.dinos)];
};

export const plugins = (opts = {}) => {
  return [dinos.plugins(opts.dinos)];
};

export const commands = {
  dinos: dinos.commands,
};
