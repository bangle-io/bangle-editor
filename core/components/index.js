// marks
export * as bold from './bold';
export * as code from './code';
export * as italic from './italic';
export * as link from './link';
export * as strike from './strike';
export * as underline from './underline';

// nodes
export * as doc from './doc';
export * as paragraph from './paragraph';
export * as text from './text';
export * as blockquote from './blockquote';
export * as bulletList from './bullet-list';
export * as codeBlock from './code-block';
export * as hardBreak from './hard-break';
export * as heading from './heading';
export * as horizontalRule from './horizontal-rule';
export * as listItem from './list-item/list-item';
export * as orderedList from './ordered-list';
export * as image from './image';

// components
export * as history from './history';
export * as editorStateCounter from './editor-state-counter';
