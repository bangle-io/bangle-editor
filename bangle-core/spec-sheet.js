import { Schema } from 'prosemirror-model';
import { coreSpec } from './components/index';
import { bangleWarn, recursiveFlat } from './utils/js-utils';

export class SpecSheet {
  constructor(rawSpecs = coreSpec()) {
    this._spec = recursiveFlat(rawSpecs);

    this._spec.forEach(validateSpec);

    if (this._spec.length !== new Set(this._spec.map((r) => r.name)).size) {
      bangleWarn(
        'The specSheet has one or more specs with the same name',
        this._spec,
      );
      throw new Error('Duplicate spec error, please check your specSheet');
    }
    this._schema = createSchema(this._spec);
    this._options = Object.fromEntries(
      this._spec
        .filter((spec) => spec.options)
        .map((spec) => [spec.name, spec.options]),
    );
  }

  get spec() {
    return this._spec;
  }

  get schema() {
    return this._schema;
  }

  get options() {
    return this._options;
  }
}

function createSchema(specSheet) {
  let nodes = [];
  let marks = [];
  let topNode;
  for (const spec of specSheet) {
    if (spec.type === 'node') {
      nodes.push([spec.name, spec.schema]);
    } else if (spec.type === 'mark') {
      marks.push([spec.name, spec.schema]);
    } else if (spec.type === 'component') {
    } else {
      throw new Error(spec.name + ' unknown type: ' + spec.type);
    }
    if (spec.topNode === true) {
      topNode = spec.name;
    }
  }

  return new Schema({
    topNode,
    nodes: Object.fromEntries(nodes),
    marks: Object.fromEntries(marks),
  });
}

function validateSpec(spec) {
  if (!spec.name) {
    bangleWarn("The spec didn't have a name field", spec);
    throw new Error('Spec must have a name');
  }
  if (!['node', 'mark', 'component'].includes(spec.type)) {
    bangleWarn('The spec must be of type node, mark or component ', spec);
    throw new Error('Invalid spec type');
  }
  if (['node', 'mark'].includes(spec.type) && !spec.schema) {
    bangleWarn(
      "The spec of type 'mark' or 'node' must have a schema field",
      spec,
    );
    throw new Error('Invalid spec schema');
  }
}
