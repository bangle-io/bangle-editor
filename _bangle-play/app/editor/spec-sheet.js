import '@banglejs/core/style.css';
import '@banglejs/tooltip/style.css';

import * as collab from '@banglejs/collab/client/collab-extension';
import { emoji } from '@banglejs/emoji/index';
import '@banglejs/emoji/emoji.css';
import '@banglejs/react-menu/style.css';
import '@banglejs/markdown-front-matter/style.css';
import './extensions-override.css';
import { trailingNode } from '@banglejs/trailing-node';
import { timestamp } from '@banglejs/timestamp';
import { SpecRegistry } from '@banglejs/core/spec-registry';
import stopwatch from '@banglejs/react-stopwatch';
import sticker from '@banglejs/react-sticker';
import { emojiSuggestMenu } from '@banglejs/react-menu';
import { markdownFrontMatter } from '@banglejs/markdown-front-matter';
import {
  bold,
  code,
  italic,
  strike,
  link,
  underline,
  doc,
  text,
  paragraph,
  blockquote,
  bulletList,
  codeBlock,
  hardBreak,
  heading,
  horizontalRule,
  listItem,
  orderedList,
  todoItem,
  todoList,
  image,
} from '@banglejs/core/components/index';

export const specRegistry = new SpecRegistry([
  doc.spec({ content: 'frontMatter? block+' }),
  text.spec(),
  paragraph.spec(),
  blockquote.spec(),
  bulletList.spec(),
  codeBlock.spec(),
  hardBreak.spec(),
  heading.spec(),
  horizontalRule.spec(),
  listItem.spec(),
  orderedList.spec(),
  todoItem.spec(),
  todoList.spec(),
  image.spec(),
  bold.spec(),
  code.spec(),
  italic.spec(),
  strike.spec(),
  link.spec(),
  underline.spec(),
  collab.spec(),
  emoji.spec(),
  emojiSuggestMenu.spec(),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  sticker.spec(),
  markdownFrontMatter.spec(),
]);
