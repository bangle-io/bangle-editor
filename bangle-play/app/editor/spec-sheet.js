import 'bangle-core/style.css';
import { coreSpec } from 'bangle-core/components';
import * as collab from 'bangle-plugins/collab/client/collab-extension';
import { emoji, emojiInlineSuggest } from 'bangle-plugins/emoji/index';
import 'bangle-plugins/emoji/emoji.css';
// import 'bangle-react/inline-menu/inline-menu.css';
import 'bangle-react/menu/style.css';

import { config } from './config';
import { trailingNode } from 'bangle-plugins/trailing-node/index';
import { timestamp } from 'bangle-plugins/timestamp/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { dino } from 'bangle-react/dino';
import { stopwatch } from 'bangle-react/stopwatch';

export const specSheet = new SpecSheet([
  ...coreSpec({ heading: { levels: config.headingLevels } }),
  collab.spec(),
  emoji.spec(),
  emojiInlineSuggest.spec({ markName: 'emoji_inline_suggest', trigger: ':' }),
  stopwatch.spec(),
  trailingNode.spec(),
  timestamp.spec(),
  dino.spec(),
]);
