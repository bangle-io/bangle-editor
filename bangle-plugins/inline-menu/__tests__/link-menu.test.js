/**
 * @jest-environment jsdom
 */

/** @jsx psx */
import { coreSpec } from 'bangle-core/components';
import { link } from 'bangle-core/index';
import { SpecSheet } from 'bangle-core/spec-sheet';
import { psx, renderTestEditor } from 'bangle-core/test-helpers/index';
import { TextSelection } from 'prosemirror-state';
import { linkMenu } from '../index';
jest.mock('bangle-plugins/helpers/index', () => {
  return {
    viewHasFocus: () => true,
  };
});

const specSheet = new SpecSheet([...coreSpec(), linkMenu.spec()]);
const plugins = [link.plugins(), linkMenu.plugins()];

describe('Link menu', () => {
  test('when no link', async () => {
    const testEditor = renderTestEditor({ specSheet, plugins });
    const { view } = await testEditor(
      <doc>
        <para>foo[]bar</para>
      </doc>,
    );

    expect(view.dom.parentNode).toMatchSnapshot();
  });

  test('when link but not in selection', async () => {
    const testEditor = renderTestEditor({ specSheet, plugins });
    const { view } = await testEditor(
      <doc>
        <para>
          foo
          <link href="https://example.com">hello world</link>
          ba[]r
        </para>
      </doc>,
    );

    expect(view.dom).toMatchInlineSnapshot(`
      <div
        class="ProseMirror bangle-editor content"
        contenteditable="true"
      >
        <p>
          foo
          <a
            href="https://example.com"
            rel="noopener noreferrer nofollow"
          >
            hello world
          </a>
          bar
        </p>
      </div>
    `);
    expect(view.dom.parentNode).toMatchSnapshot();
  });

  test('when selection moves inside selection', async () => {
    const testEditor = renderTestEditor({ specSheet, plugins });

    const { view } = await testEditor(
      <doc>
        <para>
          foo
          <link href="https://example.com">hello world</link>
          ba[]r
        </para>
      </doc>,
    );

    let tooltipDOM = view.dom.parentElement.querySelector(
      '.bangle-tooltip[data-tooltip-name="link_menu_tooltip"]',
    );
    expect(tooltipDOM).toBeTruthy();
    expect(tooltipDOM.hasAttribute('data-show')).toBe(false);
    view.dispatch(
      view.state.tr.setSelection(
        TextSelection.create(view.state.doc, view.state.doc.content.size - 6),
      ),
    );

    expect(view.state.selection).toMatchInlineSnapshot(`
      Object {
        "anchor": 13,
        "head": 13,
        "type": "text",
      }
    `);

    expect(tooltipDOM.hasAttribute('data-show')).toBe(true);
  });
});
