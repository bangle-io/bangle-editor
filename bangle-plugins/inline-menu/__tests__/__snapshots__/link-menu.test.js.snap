// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link menu when link but not in selection 2`] = `
<div
  data-testid="test-editor"
  id="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
  >
    <p>
      foo
      <a
        href="https://example.com"
        rel="noopener noreferrer nofollow"
      >
        hello world
      </a>
      bar
    </p>
  </div>
  <div
    class="bangle-tooltip bangle-selection-tooltip"
    data-tooltip-name="link_menu_tooltip"
    role="tooltip"
  >
    <div
      class="bangle-tooltip-content"
    />
  </div>
</div>
`;

exports[`Link menu when no link 1`] = `
<div
  data-testid="test-editor"
  id="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
  >
    <p>
      foobar
    </p>
  </div>
  <div
    class="bangle-tooltip bangle-selection-tooltip"
    data-tooltip-name="link_menu_tooltip"
    role="tooltip"
  >
    <div
      class="bangle-tooltip-content"
    />
  </div>
</div>
`;
